packages/cli/src/chat/chat-service.ts

```diff
@@ -0,0 +1,298 @@
+import { Logger } from '@n8n/backend-common';
+import { Service } from '@n8n/di';
+import { jsonParse, UnexpectedError } from 'n8n-workflow';
+import { type RawData, WebSocket } from 'ws';
+
+import { ChatExecutionManager } from './chat-execution-manager';
+import type { ChatMessage, ChatRequest, Session } from './chat-service.types';
+import {
+	getLastNodeExecuted,
+	getMessage,
+	isResponseNodeMode,
+	prepareMessageFromLastNode,
+	shouldResumeImmediately,
+} from './utils';
+
+const CHECK_FOR_RESPONSE_INTERVAL = 3000;
+const DRAIN_TIMEOUT = 50;
+const HEARTBEAT_INTERVAL = 30 * 1000;
+const HEARTBEAT_TIMEOUT = 60 * 1000;
+
+function closeConnection(ws: WebSocket) {
+	if (ws.readyState !== WebSocket.OPEN) return;
+
+	ws.once('drain', () => {
+		ws.close();
+	});
+
+	setTimeout(() => {
+		if (ws.readyState === WebSocket.OPEN) {
+			ws.close();
+		}
+	}, DRAIN_TIMEOUT);
+}
+
+@Service()
+export class ChatService {
+	private readonly sessions = new Map<string, Session>();
+
+	constructor(
+		private readonly executionManager: ChatExecutionManager,
+		private readonly logger: Logger,
+	) {
+		setInterval(async () => await this.checkHeartbeats(), HEARTBEAT_INTERVAL);
```

Comment:
We need to have a shutdown method for this class so it can be shut down gracefully e.g. when the instance shuts down